File: 0131-palindrome-partitioning/0131-palindrome-partitioning.java
class Solution {

    /**
     * Using List<String> ds = new ArrayList<>(); is generally better for
     * maintainability and flexibility, following good software design principles.
     * Using ArrayList<String> ds = new ArrayList<>(); is acceptable if you are
     * certain you'll always use ArrayList and don't need flexibility.
     */
    public List<List<String>> partition(String s) {
        List<List<String>> ans = new ArrayList<>();

        solve(0, s, s.length(), ans, new ArrayList<>());
        return ans;
    }

    public boolean isPalindrome(int l, int r, String s) {

        while (l < r) {
            if (s.charAt(l) != s.charAt(r))
                return false;
            l++;
            r--;
        }

        return true;
    }

    public void solve(int ind, String s, int n, List<List<String>> ans, List<String> ds) {
        if (ind == n) {
            ans.add(new ArrayList<>(ds));
            return;

        }

        for (int i = ind; i < n; i++) {

            if (isPalindrome(ind, i, s)) {
                ds.add(s.substring(ind, i + 1));
                solve(i + 1, s, n, ans, ds);
                ds.remove(ds.size() - 1);

            }
        }

    }
}

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
public:
    bool isPalindrome(string& s, int l, int r) {

        while (l <= r) {
            if (s[l] != s[r])
                return false;
            l++;
            r--;
        }
        return true;
    }

    void solve(int ind, int n, vector<vector<string>>& res, vector<string>& ds,
               string s) {

        if (ind == n) {
            res.push_back(ds);
            return;
        }

        for (int i = ind; i < n; i++) {

            if (isPalindrome(s, ind, i)) {
                ds.push_back(s.substr(ind, i - ind + 1));
                solve(i + 1, n, res, ds, s);
                //ds.erase(ds.begin() + (ds.size() - 1));
                ds.pop_back();
            }
        }
    }

    vector<vector<string>> partition(string s) {
        vector<vector<string>> res;
        vector<string> ds;
        int n = s.length();
        solve(0, n, res, ds, s);
        return res;
    }
};